zxlive/mainwindow.py

```diff
@@ -56,451 +57,31 @@ def __init__(self) -> None:
-            rewrite_export_diagram.setCheckable(False)
-            rewrite_export_diagram.setAutoExclusive(False)
+        tab_widget = QTabWidget()
+        w.layout().addWidget(tab_widget)
+        tab_widget.currentChanged.connect(self._tab_changed)
 
-            rewrite_reset = QToolButton()
-            rewrite_reset.setText("Reset")
-            rewrite_reset.setCheckable(False)
-            rewrite_reset.setAutoExclusive(False)
-            rewrite_reset.setProperty("class", "danger")
+        graph = construct_circuit()
 
-            rewrite_mode_selection.clicked.connect(rewrite_button_clicked)
-            rewrite_fuse.clicked.connect(fuse_clicked)
-            rewrite_undo.clicked.connect(undo_clicked)
-            rewrite_redo.clicked.connect(redo_clicked)
-            rewrite_change_phase.clicked.connect(change_phase_clicked)
-            rewrite_change_color.clicked.connect(change_color_clicked)
-            rewrite_do_bialgebra.clicked.connect(bialgebra_clicked)
-            rewrite_gh_state.clicked.connect(gh_state_clicked)
-            rewrite_import_diagram.clicked.connect(import_diagram_clicked)
-            rewrite_export_diagram.clicked.connect(export_diagram_clicked)
-            rewrite_reset.clicked.connect(reset_clicked)
+        self.edit_panel = GraphEditPanel(graph)
+        tab_widget.addTab(self.edit_panel, "Edit")
 
-            rewrite_tool_bar = QToolBar("rewrite_tool_bar", self)
-            rewrite_tool_bar.setObjectName("rewrite_tool_bar")
-            rewrite_tool_bar.addWidget(rewrite_mode_selection)
-            rewrite_tool_bar.addWidget(rewrite_fuse)
-            rewrite_tool_bar.addWidget(rewrite_undo)
-            rewrite_tool_bar.addWidget(rewrite_redo)
-            rewrite_tool_bar.addWidget(rewrite_change_color)
-            rewrite_tool_bar.addWidget(rewrite_do_bialgebra)
-            rewrite_tool_bar.addWidget(rewrite_gh_state)
-            rewrite_tool_bar.addWidget(rewrite_import_diagram)
-            rewrite_tool_bar.addWidget(rewrite_export_diagram)
-            rewrite_tool_bar.addWidget(rewrite_reset)
-            return rewrite_tool_bar
+        self.proof_panel = ProofPanel(graph)
+        tab_widget.addTab(self.proof_panel, "Rewrite")
 
-        edit_tool_bar = make_edit_tool_bar()
-        self.addToolBar(Qt.LeftToolBarArea, edit_tool_bar)
+    def _tab_changed(self, new_tab: Tab):
+        # This method is also invoked on application launch, so check
+        # if the tab has actually changed
+        if self.current_tab != new_tab:
+            old_panel = self.edit_panel if self.current_tab == Tab.EditTab else self.proof_panel
+            new_panel = self.edit_panel if new_tab == Tab.EditTab else self.proof_panel
+            # TODO: Do we want to maintain node selections when switching
+            new_panel.graph_view.set_graph(old_panel.graph)
+            # TODO: For now we always invalidate the undo stack when switching
```

Comment:
Or, just not clearing the stack. Just don't let users export proofs that have been edited.

Reply:
I think right now we have separate undo stacks for editing and rewriting, and I'm not sure whether that's desirable either.
Although once this project is more mature, I can imagine that you just want to stick to rewriting once you have made your diagram, since it wouldn't make much sense to manually edit it. So you 'lock' your diagram, and then you only do rewrites that preserve the semantics.

Reply:
Yeah we should have a think about how we want to treat undo/redo between modes. At the moment my idea was to keep the edit and proof undo stacks separate so that undoing in rewrite mode doesn't change the graph in a way that's not a rewrite.

I agree UI to show the undo stack would be good, but I'd leave that to a separate PR since this one is already pretty chunky?

Reply:
Agreed to not add too much to this, as this PR also changes quite a bit of the code organisation, so it would be good to have it in, so that other people can continue to work on other features that depend on this organisation.